ladybug/legend.py

```diff
@@ -0,0 +1,877 @@
+            _l_par.number_of_segments - 1
+
+    @property
+    def is_min_default(self):
+        """Boolean noting whether the min is default.
+
+        Useful when deciding whether to override the legend to properly display an
+        ordinal dictionary.
+        """
+        return self._is_min_default
+
+    @property
+    def is_max_default(self):
+        """Boolean noting whether the max is default.
+
+        Useful when deciding whether to override the legend to properly display an
+        ordinal dictionary.
+        """
+        return self._is_max_default
+
+    @property
+    def is_legend_parameters_default(self):
+        """Boolean noting whether the legend_parameters is default."""
+        return self._is_legend_parameters_default
+
+    def duplicate(self):
+        """Return a copy of the current legend."""
+        return self.__copy__()
+
+    def to_json(self):
+        """Get legend as a dictionary."""
+        leg_par = None if self.is_legend_parameters_default is True \
+            else self.legend_parameters.to_json()
+        return {'values': self.values,
+                'legend_parameters': leg_par,
+                'is_min_default': self.is_min_default,
+                'is_max_default': self.is_max_default}
+
+    @staticmethod
+    def _frange(start, stop, step):
+        """Range function capable of yielding float values."""
+        while start < stop:
+            yield start
+            start += step
+
+    def __copy__(self):
+        _leg = Legend(self.values, self.legend_parameters)
+        _leg._is_min_default = self._is_min_default
+        _leg._is_max_default = self._is_max_default
+        return _leg
+
+    def __len__(self):
+        """Return length of values on the object."""
+        return len(self._values)
+
+    def __getitem__(self, key):
+        """Return one of the values."""
+        return self._values[key]
+
+    def __iter__(self):
+        """Iterate through the values."""
+        return iter(self._values)
+
+    def ToString(self):
+        """Overwrite .NET ToString."""
+        return self.__repr__()
+
+    def __repr__(self):
+        """Legend representation."""
+        return 'Ladybug Legend ({} values)'.format(len(self))
+
+
+class LegendParameters(object):
+    """Ladybug legend parameters.
+
+    Attributes:
+        min
+        max
+        number_of_segments
+        colors
+        continuous_colors
+        continuous_legend
+        title
+        ordinal_dictionary
+        number_decimal_places
+        include_larger_smaller
+        vertical_or_horizontal
+        base_plane
+        segment_height
+        segment_width
+        text_height
+        font
+
+        is_number_of_segments_default
+        is_title_default
+        is_base_plane_default
+        is_segment_height_default
+        is_segment_width_default
+        is_text_height_default
+    """
+
+    def __init__(self, min=None, max=None, number_of_segments=None,
```

Comment:
This is fine for this case but this is too many input arguments. Is it possible to bundle some of them? Do you think one will need to input all these values when the legend parameters is initialized or we can save some of them to be set as properties if needed. Can we set some of Boolean inputs as default and then user can overwrite them if they needed to do so?

```python
lp = LegendParameters(min, max, seg_count, colors)
lp.horizontal = True
lp.text_height = 5
```

Reply:
You can see that all of the inputs to the `LegendParameters` have defaults (specifying None or leaving them blank essentially sets a default).  Also, all of the properties have setters, meaning there is no pressure to specify anything on `__init__`.
So the code you have written there will work (with the exception that the `horizontal` property is called `vertical_or_horizontal`).
We could bundle some of the properties into another object but I couldn't think of any properties that could be done so neatly or intuitively. In most cases, it felt like it was adding extra complexity.  If you have a suggestion for how to group them, let me know.

Reply:
If it was up to me I would consider most of the inputs here secondary and made `__init__` like this:

```python
   def __init__(self, domain=None, segments_count=None, colors=None, title=None, base_plane=None):
```

Once the legend is created one can change the rest of the properties.

For grouping you can use
- domain (min, max) instead of min, max
- legend segment_size (x, y) instead of segment_height and segment_width
- font and text heights can also be grouped as text properties.

Reply:
Ok. I will implement the `__init__` like you suggest but I don't agree that grouping min and max together here is a good idea.  If they are grouped, there is no way to tell if only the min is default and not the max and vice versa.  This becomes detrimental as you get to the other objects that use LegendParameters.  The segment_size and text_height suffers from a similar situation since the default values for these are all inter-related (allowing the legend to scale as a whole by setting the segment_height).  Also grouping only two items together at once doesn't seem to change things much.  If we were to group 3 or more things together (maybe base plane, all of the interrelated widths/heights), that might make a difference.  However, doing this seems a bit arbitrary.

Reply:
Done.  I removed the ability to set many of the parameters with `__init__`  and they are now only set-able with the setters.

Reply:
Looks good. thanks.